### DISCUSSION TITLE
Help with Odin + WASM + SDL2

### DISCUSSION THREAD

**Discussion by NHDaly** [odin-lang/Odin#4097] [Q&A] [answered]
I've been trying to publish a small game written in Odin + SDL2 to the web via wasm.

It took me a while, but through scrounging around other emscripten examples on the internet, I've managed to put this together:
- code: https://github.com/NHDaly/sdl-wasm-odin/
- demo: https://nhdaly.github.io/sdl-wasm-odin/

But when I've tried adding any dynamic allocation to the project, it crashes in different ways.
I've managed to shrink it down to this simple reproducer: https://github.com/NHDaly/sdl-wasm-odin/commit/e85515ff1ff92a9a744b0a3f739a14e3eec085d3
With that commit, suddenly the game fails at runtime, with this error:
```
Aborted(Stack overflow! Stack cookie has been overwritten at 0x00000004, expected hex dwords 0x89BACDFE and 0x2135467, but received 0x00000000 0x40240000)

RuntimeError: Aborted(Stack overflow! Stack cookie has been overwritten at 0x00000004, expected hex dwords 0x89BACDFE and 0x2135467, but received 0x00000000 0x40240000)
    at abort (index.js:813:11)
    at checkStackCookie (index.js:606:5)
    at postRun (index.js:664:3)
    at doRun (index.js:9975:5)
    at run (index.js:9988:5)
    at runCaller (index.js:9908:19)
    at removeRunDependency (index.js:782:7)
    at receiveInstance (index.js:976:5)
    at receiveInstantiationResult (index.js:994:5)
```

Any ideas what's going on here? I've seen many others do odin + wasm projects, so I'm sure it's something I'm doing wrong.
Help is much appreciated! ❤️

**Comment by NHDaly** [comment]
Allocating the struct doesn't fail on its own, but setting any of its fields like this does.

I certainly don't see how a Stack overflow is possible there... I think it might just be a side-effect of the illegal access. In other variants, I have previously seen complaints about accessing a 0 field.

**Reply by NHDaly** [reply]
Oh also also, one other bit of context I just remembered:

I had to make some edits to the built-in SDL2 package in order to get it to compile for the freestanding_wasm32 target (which was the only target I could get to link into emcc). So that's why there's an `sdl2` directory there in the source code.

**Comment by laytan** [comment, marked as answer]
Freestanding targets do not have any allocator set up for you, I am guessing you aren't setting one which makes all allocations just return nil and probably crash because of it.

**Reply by NHDaly** [reply]
💡 aha, thank you! I am still very new to Odin, so i didn't realize that. Thank you! Is there information about how to set the allocator? I didn't find anything in https://odin-lang.org/docs/overview/#allocators.

I would think that reusing the main OS heap allocator would still be what we want, since emscripten is going to provide wasm-ready versions of malloc/free? (EDIT: Searching the docs I found `runtime.heap_allocator()` - also crashes, same as my next comment below 😊)
Do I just need to set that allocator into the context?
I see now this other wasm project [sets an Arena allocator in its context setup](https://github.com/Caedo/raylib_wasm_odin/blob/d4078d7c50aac13ca907bbb8bc4da655516e3bee/src/main.odin#L34-L41). 👍 Thanks again for the tip.

(Or should I not be using freestanding? I found the other js targets all failed with other errors that I didn't understand.)

**Reply by NHDaly** [reply]
🤔 Any chance you could take one more look here?:
https://github.com/NHDaly/sdl-wasm-odin/commit/c08e7173b15e8f97cd14566a57a2f54c07de40bb

I've added arena allocators to the context, but i'm still seeing the exact same crash:
```
Aborted(Stack overflow! Stack cookie has been overwritten at 0x00000004, expected hex dwords 0x89BACDFE and 0x2135467, but received 0x00000000 0x40240000)

RuntimeError: Aborted(Stack overflow! Stack cookie has been overwritten at 0x00000004, expected hex dwords 0x89BACDFE and 0x2135467, but received 0x00000000 0x40240000)
```
EDIT: I've also confirmed that yes the object returned from `new` is nil.
EDIT: And i've tried with the runtime.heap_allocator() and i get nil objects as well.

**Reply by NHDaly** [reply]
This code is enough for me to trigger the assertion:
```odin

ctx: runtime.Context

tempAllocatorData: [mem.Megabyte * 4]byte
tempAllocatorArena: mem.Arena

mainMemoryData: [mem.Megabyte * 16]byte
mainMemoryArena: mem.Arena

@(export, link_name="init_sdl")
init_sdl :: proc "c" () {
    ctx = runtime.default_context()
    context = ctx

    // Initialize the allocators used in this application.
    mem.arena_init(&mainMemoryArena, mainMemoryData[:])
    mem.arena_init(&tempAllocatorArena, tempAllocatorData[:])
    ctx.allocator      = mem.arena_allocator(&mainMemoryArena)
    ctx.temp_allocator = mem.arena_allocator(&tempAllocatorArena)

    // Test the allocator works
    x := new(i64)
    assert(x != nil)  // fails
}
```

**Reply by NHDaly** [reply]
Oh, wow. It turns out that the implicit `context` variable is a struct, not a pointer, so the issue was that I was mutating `ctx`, but not `context`!

All I'm missing is this simple line _again_ after the modifications to ctx: `context = ctx`. This finally works now!
```odin
    ctx = runtime.default_context()
    context = ctx

    // Initialize the allocators used in this application.
    mem.arena_init(&mainMemoryArena, mainMemoryData[:])
    mem.arena_init(&tempAllocatorArena, tempAllocatorData[:])
    ctx.allocator      = mem.arena_allocator(&mainMemoryArena)
    ctx.temp_allocator = mem.arena_allocator(&tempAllocatorArena)

    context = ctx

    // Test the allocator works
    x := new(i64)
    assert(x != nil)
```
thanks again
